amethyst_network/src/server/server.rs

```diff
@@ -0,0 +1,105 @@
+use super::{ReceiveHandler, SendHandler, ServerConfig, ServerSocketEvent, UdpReceiver, UdpSender};
+use laminar::{error::NetworkResult, DeliveryMethod, Packet};
+use std::net::SocketAddr;
+use std::sync::Arc;
+
+/// 'Server' abstracts TCP and UDP sockets away.
+pub struct Server {
+    config: ServerConfig,
+    // Handler to access the internals of the UDP receiving thread
+    udp_receiver: Option<ReceiveHandler>,
+    // Handler to access the internals of the TCP receiving thread
+    tcp_receiver: Option<ReceiveHandler>,
+    // Handler to access the internals of the UDP sender thread
+    udp_sender: Option<SendHandler>,
+    // Handler to access the internals of the TCP sender thread
+    tcp_sender: Option<Arc<SendHandler>>,
+}
+
+impl Server {
+    /// 'Server' abstracts TCP and UDP sockets away into an uniformed API.
+    pub fn new(config: ServerConfig) -> Self {
+        Server {
+            config: config.clone(),
+
+            udp_receiver: None,
+            tcp_receiver: None,
+
+            udp_sender: None,
+            tcp_sender: None,
+        }
+    }
+
+    /// This will start the server.
+    /// 1: Fire up a TCP-sender and TCP-receiver if enabled in config.
+    /// 2: Fire up a UDP-sender and UDP-receiver.
+    /// 3: Set up some `channels` to communicate with underlying threads.
+    ///
+    /// The method uses the config provided when creating a `Server` instance.
+    pub fn run(&mut self) {
+        if self.config.enable_tcp {
+            // todo: create TCP support.
+        }
+
+        // setup a UDP-receiver which will receive packets from any endpoint.
+        self.udp_receiver = Some(UdpReceiver::run(self.config.udp_recv_addr));
+
+        // setup the UDP-sender which will send packets to an certain endpoint.
+        self.udp_sender = Some(UdpSender::run(self.config.udp_send_addr));
+    }
+
+    /// Get the handle to the internals of the UDP-receiving threat.
+    pub fn get_udp_receive_handle(&self) -> &Option<ReceiveHandler> {
+        &self.udp_receiver
+    }
+
+    /// Get the handle to the internals of the UDP-sending thread.
+    pub fn get_udp_send_handle(&self) -> &Option<SendHandler> {
+        &self.udp_sender
+    }
+
+    /// Schedule a TCP-packet for sending.
+    pub fn send_tcp(&mut self, payload: &[u8]) -> NetworkResult<()> {
+        unimplemented!()
```

Comment:
we should try and avoid putting these into master so I think if this PR isn't going to include this we should have an issue and remove this function. aka I don't want these to slip out into an actual release.